src/main/java/org/sonar/plugins/jacoco/KotlinFileLocator.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * SonarQube JaCoCo Plugin
+ * Copyright (C) 2018-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.jacoco;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+public class KotlinFileLocator {
+  private static final Logger LOGGER = Loggers.get(KotlinFileLocator.class);
+  private static final String SHEBANG_LINE = "#!.*+\n";
+  private static final String DELIMITED_COMMENT = "/\\*[^(*/)]*+\\*/";
+  private static final String LINE_COMMENT = "//[^\n]*+(\n|$)";
+  private static final String STRING_LITERAL = "\"[^\"]*+\"";
+  private static final String MULTILINE_STRING_LITERAL = "\"\"\"(?:(?!\"\"\")(.|\r?\n))*+\"\"\"";
+
+  private static final Pattern TO_SKIP = Pattern.compile(SHEBANG_LINE + "|" + DELIMITED_COMMENT + "|" + LINE_COMMENT + "|" + STRING_LITERAL + "|" + MULTILINE_STRING_LITERAL);
+
+  private static final Pattern PACKAGE_DECLARATION_REGEX = Pattern.compile("package\\s++(?<packageName>(\\w+|`[^`]++`)(\r?\n)*+(\\s*+\\.(\\w+|`[^`]++`)((\r?\n)*+|$))*+)");
+
+
+  private final Map<String, InputFile> fqnToInputFile = new HashMap<>();
+  private boolean populated = false;
+  private final Stream<InputFile> inputFileStream;
+
+  public KotlinFileLocator(Stream<InputFile> kotlinInputFileStream) {
+    inputFileStream = kotlinInputFileStream;
+  }
+
+  public InputFile getInputFile(String packagePath, String fileName) {
+    String fqn = packagePath.replace("/", ".") + "." + fileName;
+    if(!populated) {
+      populate();
+    }
+    return fqnToInputFile.get(fqn);
+  }
+
+  private void populate() {
+    inputFileStream.forEach(f -> {
+      try {
+        String packageName = getPackage(f.contents());
+        if (packageName != null) {
+          String key = packageName + "." + f.filename();
```

Comment:
We are mapping the packageName + the name of the *source* file to the file location here. Do we know the unqualified name of the source file when we read a Jacoco report? Does this work for class `Foo` that is defined in file `Bar.kt`, that also defines class `Bar`?

Reply:
Yes